### PR TITLE
Modify getting Metaflow version

### PR DESCRIPTION
**Pull request by romain-intel**
Favors reading the INFO file if present to be able to have the most accurate version of Metaflow when executing remotely (especially in the presence of extensions).

Also limit reading the INFO file to once per process (as opposed to possibly twice).

Finally, gets the version of the source of Metaflow (and not the current directory)

### CODE REVIEW

File: metaflow/__init__.py [hunk 1]
```diff
@@ -164,8 +161,13 @@ class and related decorators.
             alias_submodules(extension_module, tl_package, None, extra_indent=True)
         )
         version_info = getattr(extension_module, "__mf_extensions__", "<unk>")
-        if extension_module.__version__:
-            version_info = "%s(%s)" % (version_info, extension_module.__version__)
+        ext_version = _format_git_describe(
+            _call_git_describe(cwd=os.path.dirname(extension_module.__file__))
+        )
+        if ext_version is None:
```
Comment (savingoyal) [inline metaflow/__init__.py:167]: shouldn't this be the other way around?

also, i am getting `Metaflow 2.12.18.post5-git2201f4e-dirty+ob(4.3.19) executing...` - the interesting thing is that we don't have version 4.3.19 for any extensions internally. we pin the extension version explicitly to v1. it's likely that `git describe --tags` is the offender here.

File: metaflow/metaflow_version.py [hunk 2]
```diff
@@ -57,87 +56,110 @@ def find_git_on_windows():
...
 
     """
 
-    version = format_git_describe(call_git_describe(), pep440=pep440)
-    version_addl = None
-    if version is None:  # not a git repository
-        import metaflow
-
+    # To get the version we do the following:
+    #  - First check if we have an INFO file with it. If so, use that as it is
```
Comment (savingoyal) [inline metaflow/metaflow_version.py:138]: what is _it_ here?

File: metaflow/metaflow_version.py [hunk 3]
```diff
@@ -57,87 +56,110 @@ def find_git_on_windows():
...
+    #    the most reliable way to get the version. In particular, when running remotely,
+    #    metaflow is installed in a directory and if any extension using distutils,
```
Comment (savingoyal) [inline metaflow/metaflow_version.py:140]: grammar nit

File: metaflow/__init__.py (hunk 1 above)
Comment (savingoyal) [reply metaflow/__init__.py:167]: @madhur-ob do you happen to know how ob releases are git tagged?

**Review by savingoyal — CHANGES_REQUESTED** [review]
seems like this change doesn't quite work for our current set of extensions. @madhur-ob can you help triage what might be going wrong?

File: metaflow/metaflow_version.py [hunk 4]
```diff
@@ -57,87 +56,110 @@ def find_git_on_windows():
...
+            # first, make sure we are actually in a Metaflow repo,
+            # not some other repo
             arguments = [GIT_COMMAND, "rev-parse", "--show-toplevel"]
             reponame = (
-                check_output(arguments, cwd=CURRENT_DIRECTORY, stderr=fnull)
-                .decode("ascii")
-                .strip()
+                check_output(arguments, cwd=cwd, stderr=DEVNULL).decode("ascii").strip()
             )
+            print("For %s got reponame: %s" % (cwd, reponame))
```
[the remark this answers is not in the document]
Comment (romain-intel) [reply metaflow/metaflow_version.py:70]: Right -- that was a debugging leftover -- I'll remove it.

File: metaflow/metaflow_version.py (hunk 2 above)
Comment (romain-intel) [reply metaflow/metaflow_version.py:138]: I'll rephrase: we have an INFO file present

File: metaflow/metaflow_version.py (hunk 3 above)
Comment (romain-intel) [reply metaflow/metaflow_version.py:140]: rephrase: if any extension is using distutils to return its version

**Review by saikonen — APPROVED** [review]

**Pull request merged**